src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -302,11 +303,42 @@ private Task WriteSingleChunk<TWriteAdapter>(TWriteAdapter writeAdapter, ReadOnl
                 return WaitForWriteIOSlot(writeAdapter, ioSlot, buffer);
             }
 
-            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length + FrameOverhead);
-            byte[] outBuffer = rentedBuffer;
+            Task semaphoreTask = writeAdapter.ThrottleAsync();
 
-            SecurityStatusPal status = _sslState.EncryptData(buffer, ref outBuffer, out int encryptedBytes);
+            if (semaphoreTask.IsCompletedSuccessfully)
+            {
+                byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(buffer.Length + FrameOverhead);
+                byte[] outBuffer = rentedBuffer;
+
+                SecurityStatusPal status = _sslState.EncryptData(buffer, ref outBuffer, out int encryptedBytes);
+                s_throttle.Release();
+                return WriteEncryptedDataAsync(writeAdapter, outBuffer, rentedBuffer, encryptedBytes, status);
+            }
+
+            return CompleteAsync(writeAdapter, semaphoreTask);
+
+            async Task CompleteAsync(TWriteAdapter wAdapter, Task sTask)
+            {
+                await semaphoreTask.ConfigureAwait(false);
```

Comment:
This incurs a closure allocation even on the fast path.

Reply:
Yes I did... rookie mistake

Reply:
Nah, it's way too easy for anyone to make that mistake.  This is an example of why I want the ability to mark lambdas and local functions as non-capturing.
cc: @MadsTorgersen

Reply:
I normally do the move it outside the outer function and look for errors but yeah it catches me a bit